### ISSUE TITLE
Continuous Integration for Windows via Appveyor

### ISSUE THREAD

**Issue by djberg96** [rubygems/rubygems#1270]
Any objection to me setting up rubygems on Appveyor? I think I can do it without any special permissions since I'm a member of this group.

For environments I was going to add the following:

* 1.9.3
* 2.0.0
* 2.0.0-x64
* 2.1.0
* 2.1.0-x64

I'll add 2.2 eventually, but I don't think they've got the latest installer going, and the current one gacks on file globbing. Once they upgrade I'll add 2.2.2 and 2.2.2-x64.

I don't think we need to test any branches or tags other than current master. If you disagree, let me know.

I have been having some timeout issues with Appveyor, but caching should help, and using powershell (instead of cmd.exe) seems to help, too.

**Comment by envygeeks** [comment]
Why 1.9.3 if it's EOL?

**Comment by simi** [comment]
Actually it [looks like](https://github.com/rubygems/rubygems/blob/9d19e39dfae577cbda3d07435ec0992edab2ff75/.travis.yml#L18-L24) rubygems still supports EOL Ruby (1.8.7 and 1.9.2, 1.9.3) versions.

**Comment by tenderlove** [comment]
@djberg96 :+1:

**Comment by djberg96** [comment]
Looks like there's already an appveyor.yml file in the repo. Didn't even notice it until now, and I don't know if Luis is running it himself or what. I'm curious why we need -rtravis, too.

**Comment by duckinator** [comment]
This seems like the best approach to CI. If yes, is there anything blocking this from being done?

**Comment by indirect** [comment]
@duckinator let's set up an Appveyor account controlled by the current maintainers team, and paid for by Ruby Together. Would you be okay doing that when you want a break from triage (or after triage is done)?

**Comment by duckinator** [comment]
@indirect yeah, that works. Also, Appyveyor is apparently free for open-source projects.

**Comment by duckinator** [comment]
Been trying to get this working with my fork of the rubygems repo.

Currently looking into resolving [the following error](https://ci.appveyor.com/project/duckinator/rubygems/build/job/hq3j48t53h513at6):

```
rake's executable "rake" conflicts with C:/Ruby200/bin/rake
Overwrite the executable? [yN]
```

**Comment by duckinator** [comment]
It looks like it's related to this bug: https://bugs.ruby-lang.org/issues/5060

I need to either:

1. Find a way to specify that it should replace all executables if there are conflicts, or
2. Find an equivalent of `yes | ...` that actually works on Windows.

I don't believe 1 exists, currently.

**Comment by duckinator** [comment]
Okay, making some solid progress on this!

[I set up Appveyor for my fork of the repo](https://ci.appveyor.com/project/duckinator/rubygems). It was [getting blocked](https://ci.appveyor.com/project/duckinator/rubygems/build/1/job/hq3j48t53h513at6) before ever trying to build anything, but I set `GEM_HOME` in appyveyor.yml (https://github.com/duckinator/rubygems/commit/b946dcfe770cedbd226e895a842824b0ef777684), and that was enough to unblock the build process.

I'm not sure what Ruby 1.9.3 is doing, but [it seems incredibly unhappy](https://ci.appveyor.com/project/duckinator/rubygems/build/job/drnk0p3t8ap890lc):

```
rake aborted! 
can't convert Hash into String 
C:/Ruby193/lib/ruby/1.9.1/rake/file_list.rb:399:in `dirname'
C:/Ruby193/lib/ruby/1.9.1/rake/file_list.rb:399:in `each_dir_parent'
C:/Ruby193/lib/ruby/1.9.1/rake/dsl_definition.rb:66:in `directory'
C:/Users/appveyor/AppData/Roaming/.gem/gems/rake-10.5.0/lib/rake/packagetask.rb:146:in `define'
C:/Ruby193/lib/ruby/site_ruby/1.9.1/rubygems/package_task.rb:99:in `define'
C:/Ruby193/lib/ruby/site_ruby/1.9.1/rubygems/package_task.rb:82:in `initialize'
C:/Users/appveyor/AppData/Roaming/.gem/gems/hoe-3.14.2/lib/hoe/package.rb:44:in `new'
C:/Users/appveyor/AppData/Roaming/.gem/gems/hoe-3.14.2/lib/hoe/package.rb:44:in `define_package_tasks'
C:/Users/appveyor/AppData/Roaming/.gem/gems/hoe-3.14.2/lib/hoe.rb:730:in `block in load_plugin_tasks'
C:/Users/appveyor/AppData/Roaming/.gem/gems/hoe-3.14.2/lib/hoe.rb:724:in `each'
C:/Users/appveyor/AppData/Roaming/.gem/gems/hoe-3.14.2/lib/hoe.rb:724:in `load_plugin_tasks'
C:/Users/appveyor/AppData/Roaming/.gem/gems/hoe-3.14.2/lib/hoe.rb:808:in `post_initialize'
C:/Users/appveyor/AppData/Roaming/.gem/gems/hoe-3.14.2/lib/hoe.rb:390:in `spec'
C:/projects/rubygems/Rakefile:33:in `<top (required)>'
C:/Ruby193/lib/ruby/1.9.1/rake/rake_module.rb:25:in `load'
C:/Ruby193/lib/ruby/1.9.1/rake/rake_module.rb:25:in `load_rakefile'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:501:in `raw_load_rakefile'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:82:in `block in load_rakefile'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:133:in `standard_exception_handling'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:81:in `load_rakefile'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:65:in `block in run'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:133:in `standard_exception_handling'
C:/Ruby193/lib/ruby/1.9.1/rake/application.rb:63:in `run'
C:/Ruby193/bin/rake:32:in `<main>'
```

All Ruby 2.x versions are failing on the same test: [`TestGemExtCmakeBuilder#test_self_build`](https://github.com/rubygems/rubygems/blob/master/test/rubygems/test_gem_ext_cmake_builder.rb#L20-L44):

```
  1) Error:
TestGemExtCmakeBuilder#test_self_build:
Gem::InstallError: Makefile not found
    C:/projects/rubygems/lib/rubygems/ext/builder.rb:31:in `make'
    C:/projects/rubygems/lib/rubygems/ext/cmake_builder.rb:12:in `build'
    C:/projects/rubygems/test/rubygems/test_gem_ext_cmake_builder.rb:33:in `block in test_self_build'
    C:/projects/rubygems/test/rubygems/test_gem_ext_cmake_builder.rb:32:in `chdir'
    C:/projects/rubygems/test/rubygems/test_gem_ext_cmake_builder.rb:32:in `test_self_build'
```


I get neither of these failures locally (Windows 8.1, Ruby 1.9.3 and 2.2.3), so I'll need to do some more digging to figure out what's up.

**Comment by duckinator** [comment]
Turns out `TestGemExtCmakeBuilder#test_self_build` was skipped because I didn't have CMake installed. I installed CMake and get the same error on Ruby 2.x.

**Comment by duckinator** [comment]
So, technically, this means that if I make a PR for https://github.com/duckinator/rubygems/tree/windows-ci and someone enables Appveyor for this repo, it'll work.

But we should probably get the tests passing first.

Going to look into fixing those in a bit; taking a break for now.

**Comment by duckinator** [comment]
PR opened, since it doesn't affect anything unless Appveyor is enabled. (Which it is not, since it would show up in the same area as the information about Travis CI does.)

Once that PR is merged and the [two issues I mentioned above](https://github.com/rubygems/rubygems/issues/1270#issuecomment-174247494) are resolved, Appveyor can be enabled.

**Comment by duckinator** [comment]
Re 1.9.3 failure: I [can't reproduce it locally](https://gist.github.com/duckinator/cb486753925ac0173165), but it looks like it's somehow passing a `Hash` to `File.dirname()`.

There is also something [generating commands](https://ci.appveyor.com/project/duckinator/rubygems/build/5/job/0g1e74fiq3m28i7c#L28) of the form `gem install --version '<VERSION>' --no-rdoc --no-ri`, but the single quotes make it fail on Windows. Using double quotes would make it work, but I can't figure out what is generating those commands.

**Comment by duckinator** [comment]
The Ruby 2.x issue is related to extensions that use CMake. The test that's failing is [TestGemExtCmakeBuilder#test_self_build](https://github.com/rubygems/rubygems/blob/master/test/rubygems/test_gem_ext_cmake_builder.rb#L20-L44). I can't manage to narrow it down farther than that, currently.


I'm going to be taking a break for a couple days, so if anyone wants to take a whack at figuring out these two failures, please do! Everything I've figured out so far is in comments on this issue, and there's no unmerged code changes related to this issue, currently. When I pick this task back up again, I'll leave another comment mentioning that I'm doing so.

**Comment by duckinator** [comment]
Opened a PR to disable Windows CI for 1.9.3 (https://github.com/rubygems/rubygems/pull/1464), and emailed Appveyor about accessing a VM for debugging.

Their response:

> Absolutely, you can access VM via RDP: http://www.appveyor.com/docs/how-to/rdp-to-build-worker

I'll look into this more either later this evening or tomorrow.

**Comment by duckinator** [comment]
I was able to reproduce the 2.x issue manually in the Appveyor VM, then locally. I'm not sure what is different locally than was when I was unable to reproduce it.

Steps to reproduce:

```ruby
gem install hoe-travis
gem install minitest -v "~> 4.7"
rake travis
```

**Comment by duckinator** [comment]
Verbose output: https://gist.github.com/duckinator/e6c0f09f9a9817892fa1

It can find neither [the C compiler](https://gist.github.com/duckinator/e6c0f09f9a9817892fa1#file-gistfile1-txt-L12-L17) nor [the C++ compiler](https://gist.github.com/duckinator/e6c0f09f9a9817892fa1#file-gistfile1-txt-L29-L34).

Setting `CC=gcc` and `CXX=g++` gets to an error that is due to it using the `nmake` Makefile generator.

I [created a patch](https://gist.github.com/duckinator/7baf37be55864a0a767d) which &mdash; when combined with setting those environment variables &mdash; makes the test pass on Windows, but fail everywhere else, by using the `MSYS Makefiles` generator.

@indirect any thoughts on what I should do here? This seems like a very questionable rabbit hole to continue going down _before_ Windows CI is in place.

**Comment by duckinator** [comment]
**TL;DR of this entire thread so far:** Windows CI for Ruby 2.x is currently blocked by one failing test due to CMake using the wrong Makefile Generator. Adding the ability to specify the CMake Makefile Generator, or skipping the test, should be sufficient for making this test pass. I am not sure what the correct way to implement that functionality would be, nor if it's a good idea to try it before Windows CI is in place.

**Comment by indirect** [comment]
@duckinator wow, thanks for figuring it out. I honestly have no idea what the implications are of using a different make in the tests; I think I'm okay with disabling that test on Windows, enabling Appveyor, and then working on that test later.

**Comment by duckinator** [comment]
@indirect once #1467 is merged, you can enable Appveyor without problem.

**Comment by duckinator** [comment]
#1467 has been merged. There's a new Windows-specific failure, however: https://ci.appveyor.com/project/duckinator/rubygems/build/18/job/a54qgtvmeals48nn#L251

The failing test was added in #1469; I'll investigate more later.

**Comment by segiddins** [comment]
@duckinator ouch -- give me a sec, I can fix that for you

**Comment by duckinator** [comment]
That failure was fixed by PR #1477 being merged, so all tests pass on Windows again. Appveyor can be enabled without issue. I don't have the right permissions on the GitHub organization to enable it myself, so someone else needs to handle that.

**Comment by indirect** [comment]
Weird failures from the travis before hook, but it's running! :scream: 

```
rake travis:before --trace
** Invoke travis:before (first_time)
** Invoke install_plugins (first_time)
** Execute install_plugins
ERROR:  While executing gem ... (Gem::Requirement::BadRequirementError)
    Illformed requirement [" --no-rdoc --no-ri"] 
ERROR:  While executing gem ... (Gem::Requirement::BadRequirementError)
    Illformed requirement [" --no-rdoc --no-ri"] 
ERROR:  While executing gem ... (Gem::Requirement::BadRequirementError)
    Illformed requirement [" --no-rdoc --no-ri"] 
could not install gem for git plugin 
```

**Comment by duckinator** [comment]
@indirect what's the URL for it? I expected it to be https://ci.appveyor.com/project/rubygems/rubygems (as the one for my fork is https://ci.appveyor.com/project/duckinator/rubygems), but I get "Project not found or access denied."

**Comment by indirect** [comment]
Apparently Appveyor doesn't support organization URLs? :sob: It's at https://ci.appveyor.com/project/indirect/rubygems, and the badge code is:

```
[![Build status](https://ci.appveyor.com/api/projects/status/v1p6u2totjux5j37?svg=true)](https://ci.appveyor.com/project/indirect/rubygems)
```

**Comment by indirect** [comment]
AHA figured out how to do it... I had to sign up for a new Appveyor account.

https://ci.appveyor.com/project/rubygems/rubygems
